pkg/kademlia/dht.go

```diff
@@ -5,37 +5,170 @@ package kademlia
+	if err != nil {
+		return []proto.Node{}, err
+	}
+	return convertNetworkNodes(nn), nil
 }
 
 // GetRoutingTable provides the routing table for the Kademlia DHT
-func (k Kademlia) GetRoutingTable(ctx context.Context) (RoutingTable, error) {
-	return RouteTable{}, nil
+func (k *Kademlia) GetRoutingTable(ctx context.Context) (RoutingTable, error) {
+	return RouteTable{
+		ht:  k.dht.HT,
+		dht: k.dht,
+	}, nil
 }
 
 // Bootstrap contacts one of a set of pre defined trusted nodes on the network and
 // begins populating the local Kademlia node
-func (k Kademlia) Bootstrap(ctx context.Context) error {
-	return nil
+func (k *Kademlia) Bootstrap(ctx context.Context) error {
+	return k.dht.Bootstrap()
 }
 
 // Ping checks that the provided node is still accessible on the network
-func (k Kademlia) Ping(ctx context.Context, node overlay.Node) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k *Kademlia) Ping(ctx context.Context, node proto.Node) (proto.Node, error) {
+	n := convertProtoNode(node)
+	ok, err := k.dht.Ping(n)
+	if err != nil {
+		return proto.Node{}, err
+	}
+	if !ok {
+		return proto.Node{}, NodeErr.New("node unavailable")
+	}
+	return node, nil
 }
 
 // FindNode looks up the provided NodeID first in the local Node, and if it is not found
 // begins searching the network for the NodeID. Returns and error if node was not found
-func (k Kademlia) FindNode(ctx context.Context, ID NodeID) (overlay.Node, error) {
-	return overlay.Node{}, nil
+func (k *Kademlia) FindNode(ctx context.Context, ID NodeID) (proto.Node, error) {
+	nodes, err := k.dht.FindNode([]byte(ID))
+	if err != nil {
+		return proto.Node{}, err
+
+	}
+
+	for _, v := range nodes {
+		if string(v.ID) == string(ID) {
+			return proto.Node{Id: string(v.ID), Address: &proto.NodeAddress{
+				Transport: defaultTransport,
+				Address:   fmt.Sprintf("%s:%d", v.IP.String(), v.Port),
+			},
+			}, nil
+		}
+	}
+	return proto.Node{}, NodeErr.New("node not found")
+}
+
+// ListenAndServe connects the kademlia node to the network and listens for incoming requests
+func (k *Kademlia) ListenAndServe() error {
+	if err := k.dht.CreateSocket(); err != nil {
+		return err
+	}
+
+	return k.dht.Listen()
+}
+
+func convertProtoNodes(n []proto.Node) []*bkad.NetworkNode {
+	nn := make([]*bkad.NetworkNode, len(n))
+	for i, v := range n {
+		nn[i] = convertProtoNode(v)
+	}
+
+	return nn
+}
+
+func convertNetworkNodes(n []*bkad.NetworkNode) []proto.Node {
+	nn := make([]proto.Node, len(n))
+	for i, v := range n {
+		nn[i] = convertNetworkNode(v)
+	}
+
+	return nn
+}
+
+func convertNetworkNode(v *bkad.NetworkNode) proto.Node {
+	return proto.Node{
+		Id:      string(v.ID),
+		Address: &proto.NodeAddress{Transport: defaultTransport, Address: fmt.Sprintf("%s:%d", v.IP.String(), v.Port)},
+	}
+}
+
+func convertProtoNode(v proto.Node) *bkad.NetworkNode {
+	ip := strings.Split(v.GetAddress().GetAddress(), ":")
+
+	if len(ip) == 1 {
+		ip = append(ip, "0")
```

Comment:
?

Reply:
you know, I have absolutely no idea why I did this.